### PR TITLE
[RVD1] Interval

### PR DESCRIPTION
**Pull request by patrick-schultz**
First step in RVD changes. Rewrites `Interval` to support endpoints that are `Row`s of different lengths. Hopefully comments and test suite are enough to make the semantics clear. If not, let me know what is unclear and I'll add documentation and/or test cases.

### CODE REVIEW

File: src/main/scala/is/hail/utils/IntervalTree.scala [hunk 1]
```diff
@@ -10,117 +10,128 @@ import scala.collection.mutable
...
+    !mayOverlap(pord, other)
 
-  def definitelyEmpty(pord: ExtendedOrdering): Boolean = {
-    if (!_emptyDefined) {
-      _empty = if (includesStart && includesEnd) pord.gt(start, end) else pord.gteq(start, end)
-      _emptyDefined = true
-    }
-    _empty
-  }
+  def definitelyEmpty(pord: ExtendedOrdering): Boolean = false
```
Comment (catoverdrive) [inline src/main/scala/is/hail/utils/IntervalTree.scala]: is this still called anywhere? can we remove it?

File: src/main/scala/is/hail/utils/IntervalTree.scala [hunk 2]
```diff
@@ -10,117 +10,128 @@ import scala.collection.mutable
...
+case class IntervalEndpoint(point: Any, sign: Int) extends Serializable {
+  require(-1 <= sign && sign <= 1)
+}
+
+/**
+  * 'Interval' has an implicit precondition that there exists a Hail type
+  * 't: Type' such that either
+  * - 't: TBaseStruct', and 't.relaxedTypeCheck(left)', 't.relaxedTypeCheck(right),
+  *   and 't.ordering.intervalEndpointOrdering.lt(left, right)', or
+  * - 't.typeCheck(left)', 't.typeCheck(right)', and 't.ordering.lt(left, right)'
```
Comment (catoverdrive) [inline src/main/scala/is/hail/utils/IntervalTree.scala]: I think that this block kind of means to me that "the interval has to have compatible points and can't be empty" and maybe that's worth stating somewhere in a more explicit but less precise way?

File: src/main/scala/is/hail/utils/IntervalTree.scala [hunk 3]
```diff
@@ -129,32 +140,24 @@ object Interval {
...
 
   def ordering(pord: ExtendedOrdering, startPrimary: Boolean): ExtendedOrdering = new ExtendedOrdering {
-    private def compareStart(x: Interval, y: Interval, missingGreatest: Boolean, next: (Interval, Interval) => Int): Int = {
-      val c = pord.compare(x.start, y.start, missingGreatest)
-      if (c != 0)
-        c
-      else if (x.includesStart != y.includesStart)
-        if (x.includesStart) -1 else 1
-      else next(x, y)
+    private def compareStart(x: Interval, y: Interval, missingGreatest: Boolean): Int = {
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply src/main/scala/is/hail/utils/IntervalTree.scala]: You're right, inlined is much clearer.

File: src/main/scala/is/hail/utils/IntervalTree.scala (hunk 1 above)
Comment (patrick-schultz) [reply src/main/scala/is/hail/utils/IntervalTree.scala]: Yes!

**Pull request merged**